File: 637.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        res = []
        level = [root]
        while root and level:
            res.append([node.val for node in level])
            temp = [(node.left, node.right) for node in level]
            level = [leaf for LR in temp for leaf in LR if leaf]
        res = [sum(item)*1.0/len(item) for item in res]
        return res

File: 637.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> res;
        queue<TreeNode*> q({root});
        while(!q.empty()){
            // vector<int> tmp;
            double t = 0.0;
            int sum1 = 0;
            for (int i = q.size();i>0;--i){
                TreeNode* node = q.front();
                q.pop();
                // tmp.push_back(node->val);
                sum1++;
                t += node->val;
                if (node->left != NULL){
                    q.push(node->left);
                }
                if (node->right != NULL){
                    q.push(node->right);
                }
            }
            res.push_back(t/sum1);
        }
        return res;
    }
};
